### PR TITLE
✨ Unique Key Refactor and Other Fixes

### PR DESCRIPTION
**Pull request by znatty22**
Small fixes/changes
- Preserve index col in transform
- Always do guided transformation, addresses #141 

Unique Key Changes
- Move unique key creation stuff from transform.py into a transform/common.py module
- Build unique keys with required and optional components
    - Before this change, unique keys would only get made if all of the columns existed. 
    - Now we can have required columns and optional columns. Only required columns are needed to make the unique key.
    - For example, to make a unique key for `diagnosis`, use the required components: `participant.id` and `diagnosis.name`. If the other optional components like `diagnosis.event_age_days` exist, then use those in the key construction too

### CODE REVIEW

File: kf_lib_data_ingest/etl/transform/transform.py [hunk 1]
```diff
@@ -107,155 +105,6 @@ def _validate_run_parameters(self, data_dict):
         except TypeError as e:
             raise InvalidIngestStageParameters from e
 
-    def _insert_unique_keys(self, df_dict):
```
Comment (znatty22) [inline kf_lib_data_ingest/etl/transform/transform.py]: This chunk was moved to `kf_lib_data_ingest.etl.transform.common`

File: kf_lib_data_ingest/etl/transform/common.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+        # identify concepts in the data. In the case of auto transform
+        # this means no ConceptNodes can be created and inserted into the
+        # ConceptGraph.
+        is_any_unique_keys = any([concept_attr_from(col) == UNIQUE_ID_ATTR
+                                  for col in df.columns])
+        if not is_any_unique_keys:
+            raise ValueError(
+                'No unique keys were created for table! There must '
+                'be at least 1 unique key column in a table. Zero unique keys '
+                'in a table means there is no way to any identify concept '
```
Comment (fiendish) [inline kf_lib_data_ingest/etl/transform/common.py]:
```suggestion
                'in a table means there is no way to identify any concept '
```

File: kf_lib_data_ingest/etl/transform/common.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    columns required to compose the unique key for that concept exist in
+    the DataFrame.
+
+    The value of a unique key will be a delimited string containing the
+    values from required unique key columns.
+
+    For example, given the mapped dataframe:
+
+        CONCEPT.PARTICIPANT.ID | CONCEPT.OUTCOME.VITAL_STATUS
+        -------------------------------------------------
```
Comment (fiendish) [inline kf_lib_data_ingest/etl/transform/common.py]:
```suggestion
        -----------------------------------------------------
```

File: kf_lib_data_ingest/etl/transform/common.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+                    ]
+                }
+            }
+
+    and the unique key delimiter: -
+
+    the output dataframe would be:
+
+        CONCEPT.PARTICIPANT.ID | CONCEPT.OUTCOME.VITAL_STATUS | CONCEPT.OUTCOME.UNIQUE_KEY # noqa E501
+        ------------------------------------------------------------------------------ # noqa E501
```
Comment (fiendish) [inline kf_lib_data_ingest/etl/transform/common.py]:
```suggestion
        ---------------------------------------------------------------------------------- # noqa E501
```

File: kf_lib_data_ingest/etl/transform/common.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+    return df
+
+
+def _unique_key_cols(concept_name, df, unique_key_composition,
+                     unique_key_cols, required_cols, optional_cols):
+    """
+    Compose the list of column names that are needed to build a unique key
+    for a particular concept.
+
+    A concept's unique key can be composed of other concept's unique keys.
```
Comment (fiendish) [inline kf_lib_data_ingest/etl/transform/common.py]:
```suggestion
    A concept's unique key can be composed of other concepts' unique keys.
```

File: kf_lib_data_ingest/etl/transform/common.py [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+    This is a recursive method that collects the required columns needed to
+    build a unique key column for a concept. If one of the columns
+    is a unique key it then the method will recurse in order to get
```
Comment (fiendish) [inline kf_lib_data_ingest/etl/transform/common.py]:
```suggestion
    is a unique key then the method will recurse in order to get
```

File: kf_lib_data_ingest/etl/transform/common.py [hunk 7]
```diff
@@ -0,0 +1,218 @@
...
+    the columns that make up that unique key.
```
Comment (fiendish) [inline kf_lib_data_ingest/etl/transform/common.py]:
```suggestion
    the columns which make up that unique key.
```

File: kf_lib_data_ingest/etl/transform/common.py [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+    the output dataframe would be:
+
+        CONCEPT.PARTICIPANT.ID | CONCEPT.OUTCOME.VITAL_STATUS | CONCEPT.OUTCOME.UNIQUE_KEY # noqa E501
+        ------------------------------------------------------------------------------ # noqa E501
+            PT1                 Deceased                        PT1-Deceased # noqa E501
+
+    :param df: the Pandas DataFrame that will be modified
+    :param unique_key_composition: a dict where a key is a standard concept
+    string and a value is a list of required columns needed to compose
+    a unique key for the concept.
```
Comment (fiendish) [inline kf_lib_data_ingest/etl/transform/common.py]:
```suggestion
    a unique key for the concept
```

File: tests/test_ingest_pipeline.py [hunk 9]
```diff
@@ -24,22 +25,45 @@ def test_ingest_cmd_missing_required_args():
     assert result.exit_code == COMMAND_LINE_ERROR_CODE
 
 
-def test_ingest():
+def test_ingest(tmpdir):
```
Comment (fiendish) [inline tests/test_ingest_pipeline.py]: local variable `tmpdir` does not appear to be used

File: tests/test_transform_stage.py [hunk 10]
```diff
@@ -125,16 +127,45 @@ def test_compound_unique_keys(transform_stage):
...
             assert df.apply(lambda row: func(row), axis=1).all()
 
         else:
             assert df[id_col].equals(df[ukey_col])
 
 
-def test_no_key_comp_defined(transform_stage):
+def test_unique_key_w_optional():
+    """
+    Test unique key construction for concept whose unique key that has both
```
Comment (fiendish) [inline tests/test_transform_stage.py]:
```suggestion
    Test unique key construction for concept whose unique key has both
```

File: tests/test_ingest_pipeline.py [hunk 11]
```diff
@@ -24,22 +25,45 @@ def test_ingest_cmd_missing_required_args():
...
-
-    # Test guided transform
     result = runner.invoke(cli.ingest, [ingest_config_path])
 
     assert result.exit_code == 0
     assert 'BEGIN data ingestion' in result.output
     assert 'END data ingestion' in result.output
+
+
+def test_ingest_no_transform_module(tmpdir):
```
Comment (fiendish) [inline tests/test_ingest_pipeline.py]: is the test only temporary?

**Pull request merged**